File: androidApp/src/main/java/com/medhelp/medhelp/ui/login/LoginPresenter.kt
package com.medhelp.medhelp.ui.login

import com.androidnetworking.error.ANError
import com.google.android.gms.tasks.OnCompleteListener
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.firebase.messaging.FirebaseMessaging
import com.medhelp.medhelp.R
import com.medhelp.medhelp.data.network.ProtectionData
import com.medhelp.medhelp.data.pref.PreferencesManager
import com.medhelp.medhelp.utils.main.NetworkUtils
import com.medhelp.medhelp.utils.timber_log.LoggingTree
import com.medhelp.newmedhelp.MUtils
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import timber.log.Timber
import com.medhelp.medhelp.data.network.NetworkManager as NM
import com.medhelp.shared.network.NetworkManager

class LoginPresenter(var mainView: LoginActivity) {
    var crashlytics = FirebaseCrashlytics.getInstance()

    val prefManager = PreferencesManager(mainView)
    val networkManager = NM(prefManager)
    val networkManager2 = NetworkManager()

    val mainScope = MainScope() //+

    fun onLoginClick(username: String?, password: String?) {
        if (username == null || username.isEmpty() || password == null || password.isEmpty()) {
            mainView.showError(R.string.empty_data)
            return
        }
        if (username.length != 10) {
            mainView.showError(R.string.invalid_data)
            return
        }
        if (NetworkUtils.isNetworkConnected(mainView)) {
            verifyUser(username, password)
        } else {
            mainView.showError(R.string.connection_error)
        }
    }

    private fun verifyUser(username: String, password: String) {
        mainView.showLoading()

        mainScope.launch {
            kotlin.runCatching { networkManager2.doLoginApiCall(ProtectionData().getSignature(mainView)!!, username, password) }
                .onSuccess {
                    if (it.response[0].apiKey != null) {
                        prefManager.usersLogin = it.response

                        //List<UserResponse> ll=getDataHelper().getUsersLogin();
                        if (mainView.isNeedSave) {
                            savePrivateData(username, password)
                        } else {
                            cleanPassword()
                        }

                        //if (getDataHelper().getCurrentUserInfo() == null || getDataHelper().getCurrentUserInfo().getApiKey() == null)
                        prefManager.currentUserInfo = it.response[0]
                        centerInfo
                    } else {
                        mainView.showError(R.string.err_authorize)
                        mainView.hideLoading()
                    }
            }.onFailure {
                    var str: String? = null
                    if (it is ANError) {
                        val anError = it
                        str = anError.errorBody
                        crashlytics.log(anError.errorDetail + "//1//" + anError.errorBody)
                        if (str != null && str.contains("vasdvasdasdvя")) {
                            val dd = ProtectionData().getSignature(mainView)
                            Timber.tag("my").e(
                                LoggingTree.getMessageForError(
                                    Throwable(
                                        "Несовпадение hashkey " + dd
                                    ), "LoginPresenter\$verifyUser"
                                )
                            )
                            //   Log.wtf("wtf", pd.getSignature(view.getContext()));
                        } else {
                            Timber.tag("my").e(
                                LoggingTree.getMessageForError(
                                    it,
                                    "LoginPresenter\$verifyUser "
                                )
                            )
                        }
                    } else {
                        Timber.tag("my").e(
                            LoggingTree.getMessageForError(
                                it,
                                "LoginPresenter\$verifyUser "
                            )
                        )
                        crashlytics.log(it.message!!)
                    }
                    val str11 = it.message
                    if (str11 != null && str != null && str.contains("Failed to connect to")) {
                        mainView.showError("Failed to connect")
                    } else {
                        mainView.showError(R.string.api_default_error)
                    }
                    if (mainView == null) {
                        return@onFailure
                    }
                    mainView.hideLoading()
            }
        }
    }

    private fun cleanPassword() {
        try {
            prefManager.deleteCurrentPassword()
        } catch (e: Exception) {
            Timber.tag("my").e(LoggingTree.getMessageForError(e, "LoginPresenter\$cleanPassword "))
        }
    }

    private fun savePrivateData(username: String, password: String) {
        try {
            prefManager.currentPassword = password
            prefManager.currentLogin = username
        } catch (e: Exception) {
            Timber.tag("my")
                .e(LoggingTree.getMessageForError(e, "LoginPresenter\$savePrivateData "))
        }
    }

    // getDataHelper().setCurrentCenterUrl(centerResponses.getResponse().get(0).getIpAddress());
    private val centerInfo: Unit
        private get() {
            mainView.showLoading()

            mainScope.launch {
                kotlin.runCatching {
                    networkManager2.getCenterApiCall(prefManager.currentUserInfo!!.idCenter.toString())
                }
                    .onSuccess {
                        prefManager.centerInfo = it.response[0]
                        allHospitalBranch
                    }.onFailure {
                        if (it is ANError) {
                            val anError = it
                            crashlytics.log(anError.errorDetail + "//2//" + anError.errorBody)
                        } else {
                            crashlytics.log(it.message!!)
                        }
                        Timber.tag("my").e(LoggingTree.getMessageForError(it, "LoginPresenter/getCenterInfo"))
                        mainView.hideLoading()
                        mainView.showError("Ошибка загрузки информации о центре")
                    }
            }
        }
    private val allHospitalBranch: Unit
        private get() {

            mainScope.launch {
                kotlin.runCatching {
                    networkManager2.getAllHospitalBranch(prefManager.currentUserInfo!!.idCenter.toString())
                }
                    .onSuccess {
                        var keyExist = false
                        var allExist = true
                        for (tmp in it.response) {
                            if (tmp.idShop == null || tmp.idShop == "") {
                                allExist = false
                            } else {
                                keyExist = true
                            }
                        }
                        if (allExist) {
                            prefManager.yandexStoreIsWorks = true
                        } else {
                            prefManager.yandexStoreIsWorks = false
                            if (keyExist) {
                                Timber.tag("my").e(
                                    LoggingTree.getMessageForError(
                                        null,
                                        "Не у всех филиалов прописан Yandex IdShop"
                                    )
                                )
                            }
                        }
                        currentUserInfo
                    }.onFailure {
                        if (it is ANError) {
                            val anError = it
                            crashlytics.log(anError.errorDetail + "//3//" + anError.errorBody)
                        } else {
                            crashlytics.log(it.message!!)
                        }
                        mainView.hideLoading()
                        mainView.showError("Ошибка загрузки информации о филиалах")
                        Timber.tag("my").e(LoggingTree.getMessageForError(it, "LoginPresenter/getAllHospitalBranch "))
                    }
            }
        }

    //createList();
    private val currentUserInfo: Unit
        private get() {
            val list = prefManager.usersLogin
            if (list == null || list.size == 0) {
                Timber.tag("my")
                    .e(LoggingTree.getMessageForError(null, "LoginPresenter\$getCenterInfo 0"))
                mainView.hideLoading()
                mainView.showError("Ошибка загрузки информации о пользователе")
            }
            for (tmp in list!!) {

                mainScope.launch {
                    kotlin.runCatching {
                        networkManager2.getCurrentUserInfoInCenter(tmp.idUser!!.toString(), tmp.idBranch!!.toString(),prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
                    }
                        .onSuccess {
                            tmp.surname = MUtils.encodeDecodeWord(
                                it.responses[0].surname!!, it.responses[0].keySurname!!
                            )
                            tmp.name = it.responses[0].name
                            tmp.patronymic = it.responses[0].patronymic
                            tmp.phone = MUtils.encodeDecodeWord(
                                it.responses[0].phone!!,
                                it.responses[0].keyPhone!!
                            )
                            tmp.birthday = it.responses[0].birthday
                            tmp.email = it.responses[0].email
                            if (prefManager.currentUserInfo!!.idUser == tmp.idUser) prefManager.currentUserInfo =
                                tmp
                            var boo = true
                            for (tm in list) {
                                if (tm.name == null) {
                                    boo = false
                                    break
                                }
                            }
                            if (boo) {
                                prefManager.usersLogin = list
                                prefManager.currentUserInfo = list[0]

                                firebaseId
                            }
                        }.onFailure {
                            if (it is ANError) {
                                val anError = it
                                crashlytics.log(anError.errorDetail + "//4//" + anError.errorBody)
                            } else {
                                crashlytics.log(it.message!!)
                            }
                            Timber.tag("my").e(LoggingTree.getMessageForError(it, "LoginPresenter\$getCurrentUserInfo"))
                            mainView.hideLoading()
                            mainView.showError("Ошибка загрузки информации о пользователе")
                        }
                }
            }
        }

    val firebaseId: Unit
        get() {
            FirebaseMessaging.getInstance().token
                .addOnCompleteListener(OnCompleteListener { task ->
                    if (!task.isSuccessful || task.result == null) {
                        Timber.tag("my")
                            .e("LoginPresenter\$getFirebaseId getInstanceId failed " + task.exception)
                        mainView.openProfileActivity()
                        mainView.closeActivity()
                        return@OnCompleteListener
                    }
                    val token = task.result
                    sendFcmToken(token)
                })
        }

    private var countFcmSend=0
    private fun sendFcmToken(token: String) {
        val list = prefManager.usersLogin
        if (list == null || list.size == 0) {
            mainView.openProfileActivity()
            mainView.closeActivity()
            return
        }

        for (i in list) {
            mainScope.launch {
                kotlin.runCatching {
                    networkManager2.sendFcmId(i.idUser.toString(), i.idBranch.toString(), token, prefManager.currentUserInfo!!.apiKey!!, prefManager.centerInfo!!.db_name!!,prefManager.currentUserInfo!!.idUser.toString(),prefManager.currentUserInfo!!.idBranch.toString())
                }
                    .onSuccess {
                        countFcmSend--
                        if(countFcmSend<=0) {
                            mainView.openProfileActivity()
                            mainView.closeActivity()
                        }

                    }.onFailure {
                        if (it is ANError) {
                            val anError = it
                            crashlytics.log(anError.errorDetail + "//7//" + anError.errorBody)
                        } else {
                            crashlytics.log(it.message!!)
                        }
                        Timber.tag("my").e(LoggingTree.getMessageForError(it, "LoginPresenter\$restorePass "))
                        mainView.hideLoading()
                        mainView.showError("Ошибка восстановления пароля")
                    }
            }
        }
    }

    fun restorePass(username: String) {
        mainView.showLoading()

        mainScope.launch {
            kotlin.runCatching {
                networkManager2.requestNewPass(username)
            }
                .onSuccess {
                    mainView.responseOnPassRequest(it.response)
                    mainView.hideLoading()
                }.onFailure {
                    if (it is ANError) {
                        val anError = it
                        crashlytics.log(anError.errorDetail + "//7//" + anError.errorBody)
                    } else {
                        crashlytics.log(it.message!!)
                    }
                    Timber.tag("my").e(
                        LoggingTree.getMessageForError(
                            it,
                            "LoginPresenter\$restorePass "
                        )
                    )
                    mainView.hideLoading()
                    mainView.showError("Ошибка восстановления пароля")
                }
        }
    }
}

File: iosApp/iosApp/views/login/LoginPresenter.swift
//
//  LoginPresenter.swift
//  iosApp
//
//  Created by Михаил Хари on 02.06.2022.
//  Copyright © 2022 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared
import FirebaseMessaging


//extension LoginUiView {
//}

class LoginPresenter : ObservableObject {
    @Published var nextPage : String = "" //переход на след страницу
    @Published var valueShowAlert: AlertAttention?
    
    @Published var showDialog: String = ""
    @Published var showDialogSupportMSg: Bool = false
    
    @Published var username: String = ""
    @Published var password: String = ""
    @Published var togleState: Bool = true
    @Published var alertLogind : StandartAlertData?
    
    let sdk: NetworkManager
    var sharePreferenses : SharedPreferenses
    let netConnection = NetMonitor.shared

    init(){
        sdk=NetworkManager()
        sharePreferenses = SharedPreferenses()
        netConnection.startMonitoring()
        
        let status = netConnection.connType
        if status == NetMonitor.ConnectionType.unknown{
            self.showAlert("Отсутствует соединение с интернетом")
        }
        
    }
    
    func updateUsernameHint(){
        let login = sharePreferenses.currentLogin
        let pass = sharePreferenses.currentPassword
        
        if(login != nil && !login!.isEmpty){
            self.username = login!
        }
        
        if(pass != nil && !pass!.isEmpty){
            self.password = pass!
        }
    }
    
    func onLoginClick(){
       // let t1 = self.username
       // let t2 = self.password
        
        if(self.username.isEmpty || self.password.isEmpty){
            showAlert("Все поля дожны быть заполнены")
            return
        }
        
        let tmp = self.username.count
        
        if(self.username.count != 10){
            showAlert("Пожалуйста введите корректные данные")
            return
        }
        
        if netConnection.connType == NetMonitor.ConnectionType.unknown{
            self.showAlert("Отсутствует соединение с интернетом")
        }else{
            verifyUser(Constants.sig, self.username, self.password)
        }
    }
    
    func verifyUser (_ sig:String,_ login:String?,_ pass:String? ){
        self.showLoading(true)
        
        sdk.doLoginApiCall(signature: sig ,username: login!, password:pass!, completionHandler: { response, error in
            if let res : UserList = response {
                if(res.response[0].login != nil ){
                    self.sharePreferenses.usersLogin=res.response
                    
                    if(self.togleState){
                        self.sharePreferenses.currentLogin = login
                        self.sharePreferenses.currentPassword = pass
                    }else{
                        self.sharePreferenses.currentPassword = nil
                    }
                    
                    self.sharePreferenses.currentUserInfo = res.response[0]
                    
                    self.updateHeaderInfo()
                }else{
                    self.showLoading(false)
                    self.showAlert("Неверное имя пользователя или пароль")
                }
                
            } else {
                self.showLoading(false)
                if let t=error{
                    LoggingTree.e("LoginPresenter/verifyUser", t)
                }
                self.showAlert("Что-то пошло не так.")
            }
        })
    }
    
    func updateHeaderInfo(){
        let idCent=String(Int.init(truncating: self.sharePreferenses.currentUserInfo!.idCenter!))
        
        sdk.getCenterApiCall(idCenter: idCent, completionHandler: { response, error in
            if let res : CenterList = response {
                self.sharePreferenses.centerInfo = res.response[0]
                self.allHospitalBranch()
            } else {
                self.showLoading(false)
                self.showAlert("Ошибка загрузки информации о центре")

                if let t=error{
                    LoggingTree.e("LoginPresenter/updateHeaderInfo", t)
                }
            }
        })
    }
    
    
    func allHospitalBranch(){
        let idCent = String(Int.init(self.sharePreferenses.currentUserInfo!.idCenter!))
        
        sdk.getAllHospitalBranch(idCenter: idCent, completionHandler: { response, error in
            if let res : SettingsAllBaranchHospitalList = response {
                self.processingYandexKey(resp: res.response)
                self.currentUserInfo()
            } else {
                self.showLoading(false)
                self.showAlert("Ошибка загрузки информации о филиалах")
                
                if let t=error{
                    LoggingTree.e("LoginPresenter/getAllHospitalBranch", t)
                    
                }
            }
        })
    }
    func processingYandexKey(resp: [SettingsAllBranchHospitalResponse]){
        var keyExit = false
        var allExist = true
        
        for i in resp {
            if i.idShop == nil || i.idShop!.isEmpty{
                allExist = false
            }else{
                keyExit = true
            }
        }
        
        if allExist{
            self.sharePreferenses.yandexStoreIsWorks = true
        }else{
            self.sharePreferenses.yandexStoreIsWorks = false
            if keyExit {
                LoggingTree.e("LoginPresenter. Не у всех филиалов прописан Yandex IdShop")
            }
        }
        
    }
    
    
    func currentUserInfo(){
        
        let userLogin = self.sharePreferenses.usersLogin
        
        
        if userLogin!.count == 0 {
            LoggingTree.e("LoginPresenter/currentUserInfo 0 , Ошибка загрузки информации о пользователе")
            self.showLoading(false)
            self.showAlert("Ошибка загрузки информации о пользователе")
            return
        }
        
        let apiKey = String.init(self.sharePreferenses.currentUserInfo!.apiKey!)
        let h_dbName = self.sharePreferenses.centerInfo!.db_name!
        let idUser=String(Int.init(truncating: self.sharePreferenses.currentUserInfo!.idUser!))
        let idBranch=String(Int.init(truncating: self.sharePreferenses.currentUserInfo!.idBranch!))
        
        for i in  userLogin! {
            let idUserItem=String(Int.init(truncating: i.idUser!))
            let idBranchItem=String(Int.init(truncating: i.idBranch!))
            
            sdk.getCurrentUserInfoInCenter(idUser: idUserItem, idBranch: idBranchItem, h_Auth: apiKey, h_dbName: h_dbName, h_idKl: idUser, h_idFilial:  idBranch, completionHandler: { response, error in
                if let res : CurrentUserInfoList = response {
                    
                    i.surname = MUtils.companion.encodeDecodeWord(word: res.responses[0].surname!, key: res.responses[0].keySurname!)
                    i.name = res.responses[0].name
                    i.patronymic = res.responses[0].patronymic
                    i.phone = MUtils.companion.encodeDecodeWord(word: res.responses[0].phone!, key: res.responses[0].keyPhone!)
                    i.birthday = res.responses[0].birthday
                    i.email = res.responses[0].email
                    
                    if self.sharePreferenses.currentUserInfo!.idUser == i.idUser {
                            self.sharePreferenses.currentUserInfo = i
                    }
                    
                    var boo = true
                    
                    for tm in userLogin! {
                        if tm.name == nil {
                            boo = false
                            break
                        }
                    }
                    
                    if boo {
                        self.sharePreferenses.usersLogin = userLogin
                        self.firebaseId()
                    }
                    
                } else {
                    if let t=error{
                        LoggingTree.e("LoginPresenter/currentUserInfo", t)
                        self.showLoading(false)
                        self.showAlert("Ошибка загрузки информации о пользователе")
                    }
                }
            })
        }
    }
    
    func firebaseId (){
        Messaging.messaging().token { token, error in
          if let error = error {
              LoggingTree.e("LoginPresenter/firebaseId", error)
              self.showNextpage()
              self.showLoading(false)
          } else if let token = token {
          
              self.sendFcmToken(token: token)
          }
        }
    }
    
    private var countFcmSend=0
    func sendFcmToken(token : String){
        let userLogin = self.sharePreferenses.usersLogin
        
        if userLogin!.count == 0 {
            self.showNextpage()
            self.showLoading(false)
            return
        }
        
        countFcmSend=userLogin!.count
        
        let apiKey = String.init(self.sharePreferenses.currentUserInfo!.apiKey!)
        let h_dbName = self.sharePreferenses.centerInfo!.db_name!
        let idUser=String(Int.init(self.sharePreferenses.currentUserInfo!.idUser!))
        let idBranch=String(Int.init(self.sharePreferenses.currentUserInfo!.idBranch!))
        
        for i in userLogin! {
            let idUserItem=String(Int.init(i.idUser!))
            let idBranchItem=String(Int.init(i.idBranch!))
            
            sdk.sendFcmId(idUser: idUserItem, idFilial: idBranchItem, idFcm: token, h_Auth: apiKey, h_dbName: h_dbName, h_idKl: idUser, h_idFilial:  idBranch, completionHandler: { response, error in
                if let res : SimpleResBooleanAsString = response {
                    self.countFcmSend -= 1
                    if self.countFcmSend <= 0 {
                        self.showNextpage()
                        self.showLoading(false)
                    }
                    
                } else {
                    if let t=error{
                        LoggingTree.e("LoginPresenter/restorePass", t)
                        self.showLoading(false)
                        self.showAlert("Произошла непредвиденная ошибка")
                    }
                }
            })
        }
    }
    
    
    func restoreAlertClick(){
//        let d = username
//        let dd = username.count
//        let dd2 = self.username.count == 10
        
        if(username.isEmpty || username.count != 10  ){
            showAlert("Проверьте правильность ввода номера")
            return
        }

        showLoading(true)

        sdk.requestNewPass(username: username, completionHandler: { response, error in
            if let res : SimpleResponseString = response {
                self.showLoading(false)

                self.alertLogind = StandartAlertData(titel: "Изменение пароля!", text: res.response!, isShowCansel: true ,someFuncOk: {() -> Void in  self.alertLogind = nil},
                                               someFuncCancel: {() -> Void in self.alertLogind = nil})
            } else {
                self.showLoading(false)
                self.showAlert("Ошибка восстановления пароля")

                if let t=error{
                    LoggingTree.e("LoginPresenter/restoreAlertClick", t)

                }
            }
        })
        
    }
    
    func sendMsgToSupport(_ email : String, _ msg : String){
        self.showLoading( true)
        
        if(self.username.count != 10){
            self.showLoading( false)
            showAlert("Телефон не указан или указан не корректно", titleM: "Ошибка!")
        }
        
        if !Utils.textFieldValidatorEmail(email){
            self.showLoading( false)
            showAlert("Email не указан или указан не корректно", titleM: "Ошибка!")
        }
        
        if(msg.count < 10){
            self.showLoading( false)
            showAlert("Сообщения нет или оно сильно короткое. Попробуйте написать более развернуто", titleM: "Ошибка!")
        }
        
        sdk.sendMsgToSupport(login: self.username, email: email, msg: msg, completionHandler: { response, error in
            if let res : SimpleResBooleanAsString = response {
                self.showLoading(false)
                LoggingTree.v("Отправлено сообщение в техподдержку:  \(self.username) \(email) \(msg)")
                self.showAlert("Ваше обращение отправлено и будет обработано в рабочие часы техподдержки. Ответ будет предоставлен на указанный адрес электронной почты.", titleM: "Отправлено успешно!")
                
                self.msgNotification()
            } else {
                self.showLoading(false)
                self.showAlert("Что-то пошло не так.")

                if let t=error{
                    LoggingTree.e("LoginPresenter/sendMsgToSupport", t)

                }
            }
        })
        
    }
    func msgNotification(){
        
    }

    
    func showAlert(_ str: String, titleM : String = "Внимание!" ){
        valueShowAlert = AlertAttention(titel: titleM, text: str)
    }
    
    func showNextpage(){
        netConnection.stopMonitoring()
        self.nextPage = "Main"
    }
    
    func showLoading(_ isShow : Bool){
        if(isShow){
            self.showDialog = "loading"
        }else{
            self.showDialog = ""
        }
    }
    
    func showMsgSupporAlert(_ isShow : Bool){
        self.showDialogSupportMSg = isShow
    }
    
    func showAlertRestorePas(){
        self.alertLogind = StandartAlertData(titel: "Вы действительно хотите восстановить пароль?", text: "Старый пароль будет заменен на новый ", isShowCansel: true ,  someFuncOk: {() -> Void in
            self.restoreAlertClick()
            self.alertLogind = nil
            
        }, someFuncCancel: {() -> Void in self.alertLogind = nil})
    }
}

